### PR TITLE
Testing controllers with Jest

### PR DESCRIPTION
**Pull request by vketrys**
- [x] Writing success/fail test cases for each CRUD operation controller
- [x] Setting code coverage up to 90% lines
- [x] Jest investigation
- [x] Firebase Emulators Suite investigation 
- [x] Setting up testing environment
- [x] Setting up Git Hooks with Husky

Test coverage:
Снимок экрана от 2023-03-10 17-03-45

### CODE REVIEW

File: .husky/pre-push [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npm run test:coverage
```
Comment (EugeneSinitsa) [inline .husky/pre-push:4]: <3

File: firestore-debug.log [hunk 2]
```diff
@@ -0,0 +1,21 @@
+Mar 10, 2023 4:04:31 PM com.google.cloud.datastore.emulator.firestore.websocket.WebSocketServer start
```
Comment (EugeneSinitsa) [inline firestore-debug.log]: remove

File: package.json [hunk 3]
```diff
@@ -7,14 +7,18 @@
...
   "scripts": {
     "postinstall": "tsc",
     "start": "node dist/index.js",
+    "start:test": "cross-env NODE_ENV=test npm start",
     "watch-ts": "tsc -w",
     "watch-node": "nodemon ./dist/index.js",
     "lint": "eslint . --ext .ts",
-    "format": "eslint . --ext .ts --fix"
+    "format": "eslint . --ext .ts --fix",
+    "test": "npm run start:test && firebase emulators:exec 'npx jest --detectOpenHandles'",
```
Comment (EugeneSinitsa) [inline package.json:19]: 👌

File: src/constants/URL.ts [hunk 4]
```diff
@@ -7,9 +7,11 @@ export const URL = {
 	USERS: {
 		ROOT: '/users',
 		PARAMS: '/users/:id',
+		TEST: '/api/users',
 	},
 	MEMES: {
 		ROOT: '/memes',
 		PARAMS: '/memes/:memeId',
+		TEST: '/api/memes',
```
Comment (EugeneSinitsa) [inline src/constants/URL.ts]: 🍆 WTФ?

File: src/constants/codes.ts [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
 export const statusCodes = {
-	unprocessableEntity: 422,
-	forbidden: 403,
-	unauthorized: 401,
-	badRequest: 400,
-	ok: 200,
-	created: 201,
-	noContent: 204,
-	internalServerError: 500,
+	invalidToken_498: 498,
```
Comment (EugeneSinitsa) [inline src/constants/codes.ts]: 🍆

File: src/controllers/authControllers.ts [hunk 6]
```diff
@@ -1,79 +1,76 @@
...
+	const { displayName, email, password } = req.body;
 
-		//TODO: setting admin email from Firebase 
-		const role = email === process.env.EXAMPLE_EMAIL ? Roles.admin : Roles.user;
-  
-		if (!email || !password) {
-			return res.status(statusCodes.unprocessableEntity).send(!email ? responses.emailRequired : responses.passwordRequired);
-		}
+	//TODO: setting admin email from Firebase 
+	const role = email === process.env.EXAMPLE_EMAIL || email === 'test@gmail.com'
```
Comment (EugeneSinitsa) [inline src/controllers/authControllers.ts]: move to .env.test

File: src/controllers/authControllers.ts [hunk 7]
```diff
@@ -1,79 +1,76 @@
...
+import { Roles } from '../constants/roles';
+import { auth, db } from '../index';
+import Collections from '../constants/collections';
 
 dotenv.config();
 
 export const signup = async(req: Request, res: Response): Promise<Response> => {
-	try {
-		const { displayName, email, password } = req.body;
+	const { displayName, email, password } = req.body;
```
Comment (EugeneSinitsa) [inline src/controllers/authControllers.ts]: `const { displayName = '', email, password } = req.body;`

File: src/functions/isYoursPost.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import { Request, Response, NextFunction } from 'express';
-import { statusCodes } from '../constants/codes.js';
-import { db } from '../index.js';
-import Collections from '../constants/collections.js';
+import { db } from '../index';
+import Collections from '../constants/collections';
+import handleError from '../utils/handleError';
 
 export const isYoursPost = async(req: Request, res: Response, next: NextFunction): Promise<Response> => {
```
Comment (EugeneSinitsa) [inline src/functions/isYoursPost.ts]: rename

Comment (EugeneSinitsa) [reply src/functions/isYoursPost.ts]: 🍆

File: src/index.ts [hunk 9]
```diff
@@ -1,29 +1,39 @@
...
 dotenv.config();
 
 admin.initializeApp();
 
 export const db = getFirestore();
 export const storage = getStorage();
+export const auth = getAuth(firebaseApp);
+
+if (process.env.NODE_ENV === 'test') {
+	connectAuthEmulator(auth, 'http://127.0.0.1:9099', { disableWarnings: true });
```
Comment (EugeneSinitsa) [inline src/index.ts]: 🍆

File: src/index.ts [hunk 10]
```diff
@@ -1,29 +1,39 @@
...
 
 app.use(URL.ROOT, authRoutes, memeRoutes);
 
-
-app.listen(port, () => {
-	process.stdout.write(`app listening on port ${port}`);
-});
+if (process.env.NODE_ENV !== 'test') {
+	app.listen(port, () => {
+		process.stdout.write(`app listening on port ${port}`);
```
Comment (EugeneSinitsa) [inline src/index.ts]: ?

File: .env.test [hunk 11]
```diff
@@ -0,0 +1,8 @@
+PORT=3001
```
Comment (slamser56) [inline .env.test]: I guess you should remove this file according .gitignore

File: src/constants/URL.ts [hunk 12]
```diff
@@ -7,9 +7,11 @@ export const URL = {
 	USERS: {
 		ROOT: '/users',
 		PARAMS: '/users/:id',
+		TEST: '/api/users',
```
Comment (slamser56) [inline src/constants/URL.ts]: What do you use this route for?

File: src/constants/codes.ts [hunk 13]
```diff
@@ -1,15 +1,17 @@
 export const errorCodes = {
 	weakPassword: 'auth/weak-password',
 	wrongPassword: 'auth/wrong-password',
+	wrongUserID: 'auth/user-not-found',
```
Comment (slamser56) [inline src/constants/codes.ts:4]: Why do you use various styles of const?

File: src/functions/isUserId.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { Request, Response, NextFunction } from 'express';
+import { db } from '../index';
+import Collections from '../constants/collections';
+import handleError from '../utils/handleError';
+
+export const isUserId = async(req: Request, res: Response, next: NextFunction): Promise<Response> => {
```
Comment (slamser56) [inline src/functions/isUserId.ts:6]: Is it utils using for check existence of user?

File: ui-debug.log [hunk 15]
```diff
@@ -0,0 +1,2 @@
+Web / API server started at 127.0.0.1:4000
```
Comment (slamser56) [inline ui-debug.log]: I guess we don't need to push logs to git

File: src/functions/isUserId.ts (hunk 14 above)
Comment (vketrys) [reply src/functions/isUserId.ts:6]: no, it's basically checking user id

File: src/constants/codes.ts (hunk 13 above)
Comment (vketrys) [reply src/constants/codes.ts:4]: for better readability

**Review by esnts — APPROVED** [review]

**Pull request merged**